### DISCUSSION TITLE
Prevent Endpoint Listing for Next.js API routes

### DISCUSSION THREAD

**Discussion by shahruz** [vercel/vercel#4168] [General] [answered]
I couldn't find an answer to this elsewhere and it seems more of a Vercel / Now question than a Next.js question, so hopefully this is the right place to ask...

The Vercel docs show the ability to exclude files and directories in the `/api` directory starting with an `_` from generating serverless functions. 

[https://vercel.com/docs/v2/serverless-functions/introduction#prevent-endpoint-listing](https://vercel.com/docs/v2/serverless-functions/introduction#prevent-endpoint-listing)

I'm using Next.js + Vercel for a project and currently have a handful of helper functions 
like `_db.js` inside `/pages/api`. On deploy Vercel currently makes a serverless function for every file in `/pages/api` including all the underscored helper functions.

<img width="458" alt="Screen Shot 2020-04-27 at 11 56 25 AM" src="https://user-images.githubusercontent.com/573917/80409802-57e36a80-887e-11ea-8292-a6c7aa237f12.png">

Is there any way to apply the Vercel functionality to the Next.js api directory? Initially I thought customizing the `builds` or `functions` configs would help, but I can't seem to figure out a non-deprecated way to exclude those files. 

It's trivial to move all the helper files out into the root directory, but I would definitely prefer a way to keep all my API-specific code inside the `/pages/api` directory if possible.

**Comment by coetry** [comment, marked as answer]
Hi @shahruz this is actually more of a Next.js question and the Next.js core team has taken a position to prevent supporting this because it poses a security issue where users can forget to add a prefix and result in unintended code being present in a server environment.

As you said, the easiest and recommended way to handle this would be to move your helper functions out to a `/utils` or `/lib` folder (name is arbitrary) in your root. You can group API related helpers in a subdirectory like `/utils/api` for example.

**Reply by shahruz** [reply]
Cool, thanks for the clarification!

**Comment by Mihai-github** [comment]
Hi, 

   Is there a way to turn off (not sure if this is the correct way of saying this) some API Route from being serverless but still working as a node.js endpoint like
   
   You have a route called /api/upload-image/index.ts -> SERVERLESS
   To move this function or turn off Vercel from making it a serverless function.

   Why I'm asking because of the Payload Size limit of 4.5MB and seems like there is no bypass from what the docs stay but this means for example with images uploading of large sizes you have to make every upload with `presigned URLs` from the client.
